### PR TITLE
[Codegen] Support scalability when hoisting (and checking) allocations

### PR DESCRIPTION
**Pull request by MacDue**
This makes use of recent work (https://github.com/llvm/llvm-project/pull/83876) to support scalability in value bounds analysis. With these changes scalable allocations (i.e. allocations that scale with vscale), can now be hoisted and pass stack limit checking.

Hoisting scalable allocations requires knowledge of the range of vscale, which is taken from the target (using the maximum range defined by the ISA). The stack limit checking just assumes `vscale = 1` (by default), since this avoids needing to look at the target (and always resolves to a constant bound), and still prevents excessive allocations.

ci-extra: build_test_all_arm64

### CODE REVIEW

**Comment by MacDue** [comment]
~~Note: This won't build until https://github.com/llvm/llvm-project/pull/83876 is integrated into IREE's LLVM fork.~~

File: compiler/src/iree/compiler/Codegen/Common/Passes.td [hunk 1]
```diff
@@ -312,6 +312,17 @@ def HoistStaticallyBoundAllocations :
...
   let summary = "Hoist statically bound alloca ops to the entry block of functions";
   let constructor = "mlir::iree_compiler::createHoistStaticallyBoundAllocationsPass()";
+  let options = [
+    Option<"vscaleMin", "vscale-min", "unsigned",
+            /*default=*/"0",
+           "Minimum possible value of vscale.">,
+    Option<"vscaleMax", "vscale-max", "unsigned",
+            /*default=*/"0",
+           "Maximum possible value of vscale (a value of zero means unbounded).">
+  ];
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/Passes.td:324]: 1. What's the meaning of these in the case of fixed-width vectors? Could you add a comment?
2. Using `0` for `unbounded` (which happens to be the default) is a bit counter-intuitive for me. Would it be possible to test such case?

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp [hunk 2]
```diff
@@ -121,44 +124,96 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
     return allocation;
   }
 
+  Value vscale = nullptr;
+  if (!vscaleRange.has_value()) {
+    auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(funcOp);
+    vscaleRange = getDefaultVscaleRange(targetAttr);
+  }
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:132]: This logic means that the value specified by the user (when configuring `HoistStaticallyBoundAllocations`) will take precedence over function attributes. This makes sense to me, but please add a comment and test this behaviour :)

File: compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir [hunk 3]
```diff
@@ -157,3 +157,81 @@ func.func @nested_op_alloc_linalg_use(%arg0 : index) {
...
+//  CHECK-NEXT:   %[[VSCALE:.+]] = vector.vscale
+//  CHECK-NEXT:   %[[C32_VSCALE:.+]] = affine.apply #[[$SCALABLE_BOUND_MAP_2]]()[%[[VSCALE]]]
+//  CHECK-NEXT:   %[[ALLOC:.+]] = memref.alloc(%[[C32_VSCALE]]) : memref<?xi32>
+//       CHECK:   scf.for
+//       CHECK:     %[[SIZE:.+]] = affine.min
+//       CHECK:     %[[SUBVIEW:.+]] = memref.subview %[[ALLOC]][0] [%[[SIZE]]] [1]
+//       CHECK:     %[[CAST:.+]] = memref.cast %[[SUBVIEW]]
+//       CHECK:     linalg.fill
+//  CHECK-SAME:         outs(%[[CAST]] :
+//       CHECK:   memref.dealloc %[[ALLOC:.+]] : memref<?xi32>
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:249]: I know that this is merely adapting an existing test rather than something specific to your PR, but isn't hoisting in this case incorrect? As in, if `memref.alloc` is hoisted then shouldn't `memref.dealloc` be hoisted as well?

File: compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// RUN: iree-opt --split-input-file --pass-pipeline="builtin.module(func.func(iree-hoist-statically-bound-allocations))" %s | FileCheck %s
+// RUN: iree-opt --split-input-file --pass-pipeline="builtin.module(func.func(iree-hoist-statically-bound-allocations{vscale-min=1 vscale-max=16}))" %s | FileCheck %s
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:1]: Perhaps use values other then default for vscale range?

File: compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:1]: To me it would make sense to "duplicate" most/all tests with a `vscale` variant. Basically, if `non_entry_bb_scalable_allocas` is next to `@non_entry_bb_allocas` then it becomes pretty obvious that the former is just a "scalable" variant of the latter. Otherwise, these two tests are quite unrelated. Not sure we need to duplicate every test though 🤔

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp [hunk 5]
```diff
@@ -58,11 +59,14 @@ checkStackAllocationSize(mlir::FunctionOpInterface funcOp) {
       allocaSize *= dimSize;
     }
     for (auto operand : allocaOp.getDynamicSizes()) {
-      auto ub = ValueBoundsConstraintSet::computeConstantBound(
-          presburger::BoundType::UB, operand, /*dim=*/std::nullopt,
-          /*stopCondition=*/nullptr, /*closedUB=*/true);
-      if (succeeded(ub)) {
-        allocaSize *= ub.value();
+      // Assume vscale is 1 for determining if the alloca is within the stack
+      // limit. This should always resolve to a constant bound.
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: > Assume vscale is 1 for determining if the alloca is within the stack limit.

This feels wrong :) Could you explain what's happening here? In particular, why bother with "vscale range" in various places if such an assumption is being made here?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/check_ir_before_llvm_conversion_not_fail_unbound.mlir [hunk 6]
```diff
@@ -29,3 +29,16 @@ module {
...
 }
 // CHECK-LABEL: func @mix_static_and_unbound_dynamic_allocas(
+
+// -----
+
+module {
+  func.func @scalable_alloca() {
+    %c4 = arith.constant 4 : index
+    %vscale = vector.vscale
+    %c4_vscale = arith.muli %vscale, %c4 : index
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/check_ir_before_llvm_conversion_not_fail_unbound.mlir]: IIUC, all allocation in this file are meant to be out-of-bounds. However, `vscale * 4` feels to small to be out-of-bounds? What qualifies as "out-of-bounds"?

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 7]
```diff
@@ -1177,4 +1177,15 @@ bool hasFusedLeadingOp(linalg::LinalgOp rootOp) {
   });
 }
 
+std::optional<VscaleRange>
+getDefaultVscaleRange(IREE::HAL::ExecutableTargetAttr targetAttr) {
+  if (isAArch64(targetAttr)) {
+    // On AArch64 the vector length will always be between 128-bit and 2048-bit.
+    // This works out as a vscale range of 1 to 16.
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: Let's add a link: https://developer.arm.com/Architectures/Scalable%20Vector%20Extensions. Also, this is referring to SVE vector lengths rather than "vectors" in general. Worth claryfying.

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp [hunk 8]
```diff
@@ -121,44 +124,96 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
...
+
+    constantBound = *ub;
+    return success();
+  };
+
   /// For the dynamic but bounded case, insert an allocation of the shape of the
   /// bounds, and a subview of the required size to be used as a replacement.
-  SmallVector<int64_t> staticShape;
+  SmallVector<int64_t> allocShape;
+  SmallVector<Value> scalableSizes;
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp]: Could we have a comment explaining the different between the two?

**Review by banach-space — COMMENTED** [review]
Thanks Ben, makes sense!

I've made a few comments/suggestions, but the overall looks good.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp (hunk 5 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: This is for a static (i.e. at compile time) check on the number of bytes allocated in a function. We can't know what `vscale` will actually be here, so any value will be 'wrong'. However, assuming `vscale` is 1, will still catch excessive allocations (which is what I think the intention of this check is), without being overly conservative.

This does not change codegen, it's simply a check (does this function allocate too many bytes yes/no?).

File: compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir (hunk 3 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:249]: The `memref.dealloc` is hoisted (though this test would not catch if it wasn't).

Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:249]: It should be possible to update the test to capture that, right?

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:249]: Yep (I was just copy/pasting the previous test case when adding this).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp (hunk 5 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: Ah, now I see, thanks for the explanation!

Would it be possible to grab the actual values though? I guess that it can only come from a function attribute and/or target info? Otherwise I'm worried that this isn't conservative enough 🤔

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: As far as I understand it this is just a check for unreasonably large allocations, but nothing would break if at some `vscale` it happened to go over it a bit. In the size of allocations `vscale` will a small factor (with the most of the size likely from static dimensions), so I think you can get away with mostly ignoring it (and still have a reasonable check).

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: Assuming `vscale = 16` (on Arm) would be assuming a value of vscale that is very unlikely to actually occur.

File: compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir (hunk 4 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:1]: Why? This is just testing hoisting allocas, so any reasonable range should work -- with the same result (this is not testing computing the scalable bounds).

File: compiler/src/iree/compiler/Codegen/Common/Passes.td (hunk 1 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/Passes.td:324]: > Using 0 for unbounded (which happens to be the default) is a bit counter-intuitive for me. Would it be possible to test such case?

It's the pretty much the same as vscale_range is specified (in the LangRef). Unbounded is the default behaviour because that's saying nothing is known about vscale.

File: compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/test/hoist_statically_bound_allocations.mlir:1]: None of the logic I touched looks at users alloca/alloc being hoisted, so I don't think it's necessary to duplicate every test. I just chose a few representative examples. I'll move the tests though :+1:

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.h [hunk 9]
```diff
@@ -52,7 +52,8 @@ template <typename AllocLikeOpType>
 std::optional<Value> hoistOneStaticallyBoundAllocation(
     mlir::FunctionOpInterface funcOp, OpBuilder &builder, Location loc,
     MemRefType allocaType, ValueRange dynamicSizes,
-    std::optional<uint64_t> alignment);
+    std::optional<uint64_t> alignment,
+    std::optional<VscaleRange> vscaleRange = {});
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.h]: nit: it looks better if the default value is `std::nullopt`, but not `{}`. Same for below changes

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 10]
```diff
@@ -1177,4 +1177,16 @@ bool hasFusedLeadingOp(linalg::LinalgOp rootOp) {
   });
 }
 
+std::optional<VscaleRange>
+getDefaultVscaleRange(IREE::HAL::ExecutableTargetAttr targetAttr) {
+  if (isAArch64(targetAttr)) {
+    // On AArch64 the scalable vector length will always be between 128-bit and
+    // 2048-bit. This works out as a vscale range of 1 to 16. See:
+    // https://developer.arm.com/Architectures/Scalable%20Vector%20Extensions
+    return VscaleRange{1, 16};
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:1170]: (Every time I see this, I'd like to suggest to use designated initializers. Then I realized that it is a c++20 feature. :p)

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp [hunk 11]
```diff
@@ -121,44 +123,101 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
...
+  if (!vscaleRange.has_value()) {
+    auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(funcOp);
+    vscaleRange = getDefaultVscaleRange(targetAttr);
+  }
+  auto computeAllocationBound = [&](Value value, Value &scalableBound,
+                                    int64_t &constantBound) -> LogicalResult {
+    scalableBound = nullptr;
+    if (vscaleRange.has_value()) {
+      // Scalability supported: Allocations could be scalable.
+      auto ub = vector::ScalableValueBoundsConstraintSet::computeScalableBound(
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp]: Please spell the type (i.e., do not use auto) because:

1.  It is not trivial.
2. We do access a member of the variable.

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp [hunk 12]
```diff
@@ -121,44 +123,101 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
...
+      }
+
+      if (scalableBound) {
+        // Scalable size.
+        allocShape.push_back(ShapedType::kDynamic);
+        scalableSizes.push_back(scalableBound);
+      } else {
+        // Static size.
+        allocShape.push_back(constantBound);
+      }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp]: I think both alloc and alloca op take the below build method. Is it possible to let `computeAllocationBound` return `std::optional<OpFoldResult>` and use it in allocation shape?

```
static void build(::mlir::OpBuilder &odsBuilder, ::mlir::OperationState &odsState, ArrayRef<OpFoldResult> sizes, Type elementType, Attribute memorySpace = {})
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp [hunk 13]
```diff
@@ -58,11 +59,14 @@ checkStackAllocationSize(mlir::FunctionOpInterface funcOp) {
...
+      auto ub = vector::ScalableValueBoundsConstraintSet::computeScalableBound(
+          operand, /*dim=*/std::nullopt,
+          /*vscaleMin=*/1, /*vscaleMax=*/1, presburger::BoundType::UB);
+      FailureOr<vector::ConstantOrScalableBound::BoundSize> size;
+      if (succeeded(ub) && succeeded(size = ub->getSize())) {
+        allocaSize *= size->baseSize;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: I don't get why we need `size`. Can't this just be below?

```
      if (succeeded(ub)) {
        allocaSize *= ub->getSize()->baseSize;
        continue;
      }
```

File: compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp [hunk 14]
```diff
@@ -28,8 +28,17 @@ struct HoistStaticallyBoundAllocationsPass
...
   auto funcOp = getOperation();
   IRRewriter rewriter(funcOp->getContext());
-  hoistStaticallyBoundAllocationsInFunc<memref::AllocaOp>(rewriter, funcOp);
-  hoistStaticallyBoundAllocationsInFunc<memref::AllocOp>(rewriter, funcOp);
+
+  auto vscaleRange = [&]() -> std::optional<VscaleRange> {
+    if (this->vscaleMax == 0 || this->vscaleMin > this->vscaleMax)
+      return {};
+    return VscaleRange{this->vscaleMin, this->vscaleMax};
+  }();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp]: Why do we need the lambda? I think we can simplify it to

```cpp
std::optional<VscaleRange> vscaleRange;
if (this->vscaleMax != 0 && this->vscaleMin <= this->vscaleMax) {
  vscaleRange = {this->vscaleMin, this->vscaleMax};
}
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp (hunk 13 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: Probably overly cautious,  `getSize()` can fail if the bound is a more complex expression (in terms of vscale). However, that should never happen if vscale is a constant (which it is here). So `ub->getSize()->baseSize` should be fine :+1:.

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp (hunk 12 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp]: That's much nicer, thanks.

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp]: Seems like it may break everything too... :pensive:

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp [hunk 15]
```diff
@@ -121,44 +123,92 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
...
+      allocSizes.push_back(*ub);
+      subviewSizes.push_back(dynamicSize);
+    }
+
+    // FIXME: The `AllocLikeOp::build()` method for OpFoldResult drops the
+    // layout, so we have to resolve the static/dynamic values here.
+    SmallVector<int64_t> staticShape;
+    SmallVector<Value> dynamicSizes;
+    staticShape.reserve(allocLikeType.getRank());
+    dynamicSizes.reserve(allocLikeType.getRank());
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp]: [optional] I'd avoid using the trick here. This kind of tricks should be happening in `dispatchIndexOpFoldResults` implementation details.

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp [hunk 16]
```diff
@@ -121,44 +123,92 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
...
+      auto ub = computeAllocationBound(dynamicSize);
+      if (failed(ub))
+        return std::nullopt;
+
+      allocSizes.push_back(*ub);
+      subviewSizes.push_back(dynamicSize);
+    }
+
+    // FIXME: The `AllocLikeOp::build()` method for OpFoldResult drops the
+    // layout, so we have to resolve the static/dynamic values here.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:196]: ohh.. this is unfortunate..

File: compiler/src/iree/compiler/Codegen/Utils/Utils.h [hunk 17]
```diff
@@ -237,6 +237,14 @@ void sinkOpsInCFG(const SmallVector<Operation *> &allocs,
 // the inputs.
 bool hasFusedLeadingOp(linalg::LinalgOp rootOp);
 
+struct VscaleRange {
+  unsigned min;
+  unsigned max;
+};
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.h:243]: Would it make sense to make it has initial values (i.e., `0`) like what we have in `Passes.td`? In this case, perhaps we can avoid `std::optional`. Instead, we can pass `VscaleRange vscaleRange = VscaleRange{}` in the methods. What do you think?

**Review by hanhanW — APPROVED** [review]
LGTM, just few nits

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp (hunk 5 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUCheckIRBeforeLLVMConversion.cpp]: Yes, `vscale = 16` would be super conservative. AFAIK, the widest vectors that we have today are 512 bits (`vscale = 4`):
* https://en.wikipedia.org/wiki/Fujitsu_A64FX

However, we need to make sure that we don't underestimate too optimistically. It's tricky because this is a compile-time check and we won't know the value of `vscale` until runtime. I think that `clAssumedVscaleValue` that you added is a very good compromise for now. If we need to refine this logic then what you've added here will allow us to make an informed decision. Thanks!

**Review by banach-space — APPROVED** [review]
Thanks for addressing my comments, LGTM!

File: compiler/src/iree/compiler/Codegen/Utils/Utils.h (hunk 17 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.h:243]: I'd prefer to keep the `std::optional`. I don't like encoding the semantics in the value, it's just the option parser does not support `std::optional<unsigned>`, so I have to do that in `Passes.td`.

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp (hunk 16 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:196]: Fix this upstream?

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:196]: It should be :) 
But I'd prefer not to wait on the fix (and integration)

File: compiler/src/iree/compiler/Codegen/Utils/Utils.h (hunk 17 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.h:243]: I see, makes sense to me. Thanks!

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp (hunk 16 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:196]: Yes, let's do it in a follow-up..

**Comment by banach-space** [comment]
Btw, Ben has no commit access - I will land this for him either later today or tomorrow.

**Comment by MacDue** [comment]
Please hold of merging till I enable the extra CI checks (and those all pass) :)

(Running now)

**Comment by MacDue** [comment]
I think this is good to go now (was rebased on `main` + ran the extra aarch64 CI, all green here) :crossed_fingers:
